File: cbsd/cbslib/routes/components.py
# CBS service library - routes - components
# Copyright (C) 2025  Clyso GmbH
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.


from cbsdcore.api.responses import AvailableComponentsResponse, BaseErrorModel
from fastapi import APIRouter, HTTPException, status

from cbslib.builds.mgr import NotAvailableError
from cbslib.routes import logger as parent_logger
from cbslib.routes._utils import CBSAuthUser, CBSBuildsMgr, responses_auth

logger = parent_logger.getChild("components")

router = APIRouter(prefix="/components")


@router.get(
    "/",
    summary="Obtain known build components",
    responses={
        **responses_auth,
        500: {
            "description": "Unknown error obtaining components",
            "model": BaseErrorModel,
        },
        200: {
            "description": "Known components",
            "model": AvailableComponentsResponse,
        },
    },
)
async def components_list(
    user: CBSAuthUser,
    mgr: CBSBuildsMgr,
) -> AvailableComponentsResponse:
    """Obtain known components that are able to be built by the service."""
    logger.debug(f"obtain components list, user: {user}")

    try:
        return mgr.components
    except NotAvailableError:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="try again later"
        ) from None
    except Exception as e:
        logger.error(f"unknown error obtaining components: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="check logs for failure",
        ) from e

File: cbsd-rs/cbsd-server/src/routes/components.rs
// Copyright (C) 2026  Clyso
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.

//! Route handlers for `/api/components/*`: component discovery.

use axum::Json;
use axum::extract::State;
use axum::http::StatusCode;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;

use crate::app::AppState;
use crate::auth::extractors::{AuthUser, ErrorDetail};
use crate::components::ComponentInfo;

/// Build the components sub-router: `/api/components/*`.
pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new().routes(routes!(list_components))
}

// ---------------------------------------------------------------------------
// GET /api/components/
// ---------------------------------------------------------------------------

/// List all known components. Requires authentication but no specific
/// capability.
#[utoipa::path(
    get,
    path = "",
    tag = "components",
    security(("bearer" = []), ("cookie" = [])),
    responses(
        (status = StatusCode::OK, body = Vec<ComponentInfo>),
        (status = StatusCode::UNAUTHORIZED, body = ErrorDetail),
    ),
)]
async fn list_components(
    State(state): State<AppState>,
    _user: AuthUser,
) -> Result<Json<Vec<ComponentInfo>>, (StatusCode, Json<ErrorDetail>)> {
    Ok(Json(state.components.clone()))
}
